### PR TITLE
Use enums for SAS permissions

### LINKED ISSUE #6902: [P1] Storage: Use flags enums for various options types
A lot of our options types should be Flags enums instead of structs full of bools.  We can just add `To***`/`Parse` extension methods on `BlobExtensions` to serialize/deserialize as appropriate.  We'll probably need to wait for the API review to get the full list of types we should convert into enums.

Look at [`BlobSasPermissions` as an example](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/storage/Azure.Storage.Blobs/src/Sas/BlobSasPermissions.cs).  I think we make that a `[Flags] public enum BlobSasPermissions { /* ...  */ }` and add `string ToPermissionsString(this BlobSasPermissions permissions)` and `BlobSasPermissions ParsePermissionsString(string permissions)` on `BlobExtensions`, but the API review should make that clear.

### PR DESCRIPTION
**Pull request by JoshLove-msft**
Fixes #6902

### CODE REVIEW

File: sdk/storage/Azure.Storage.Blobs/src/Sas/BlobSasBuilder.cs [hunk 1]
```diff
@@ -361,6 +361,10 @@ private static string GetCanonicalName(string account, string containerName, str
         /// </summary>
         private void EnsureState()
         {
+            if (ExpiresOn == default || string.IsNullOrEmpty(Permissions))
+            {
+                throw Errors.SasMissingData();
```
Comment (tg-msft) [inline sdk/storage/Azure.Storage.Blobs/src/Sas/BlobSasBuilder.cs]: Please throw separate errors.  It's a lot less frustrating for users.  Use `nameof(ExpiresOn)`, etc., as well so refactoring keeps things updated.

File: sdk/storage/Azure.Storage.Blobs/src/Sas/BlobSasBuilder.cs [hunk 2]
```diff
@@ -52,11 +53,11 @@ public class BlobSasBuilder
...
         /// field must be omitted if it has been specified in an associated
         /// stored access policy.  The <see cref="BlobSasPermissions"/>,
-        /// <see cref="BlobContainerSasPermissions"/>, and
-        /// <see cref="SnapshotSasPermissions"/> can be used to create the
+        /// <see cref="BlobContainerSasPermissions"/>, <see cref="SnapshotSasPermissions"/>,
+        /// or <see cref="BlobAccountSasPermissions"/> can be used to create the
         /// permissions string.
         /// </summary>
-        public string Permissions { get; set; }
+        private string Permissions { get; set; }
```
Comment (tg-msft) [inline sdk/storage/Azure.Storage.Blobs/src/Sas/BlobSasBuilder.cs]: Let's make this a `public` property with a `private set`.

File: sdk/storage/Azure.Storage.Blobs/tests/BlobTestBase.cs [hunk 3]
```diff
@@ -246,75 +246,109 @@ public StorageSharedKeyCredential GetNewSharedKeyCredentials()
...
                 Services = AccountSasServices.Blobs,
                 ResourceTypes = AccountSasResourceTypes.Container | AccountSasResourceTypes.Object,
                 StartsOn = Recording.UtcNow.AddHours(-1),
                 ExpiresOn = Recording.UtcNow.AddHours(+1),
-                Permissions = new BlobContainerSasPermissions { Read = true, Add = true, Create = true, Write = true, Delete = true, List = true }.ToString(),
                 IPRange = new SasIPRange(IPAddress.None, IPAddress.None)
-            }.ToSasQueryParameters(sharedKeyCredentials);
+            };
+            builder.SetPermissions(
+                AccountSasPermissions.Read |
```
Comment (tg-msft) [inline sdk/storage/Azure.Storage.Blobs/tests/BlobTestBase.cs]: This was BlobContainerSasPermissions and now you've got AccountSasPermissions?

File: sdk/storage/Azure.Storage.Blobs/tests/BlobTestBase.cs [hunk 4]
```diff
@@ -246,75 +246,109 @@ public StorageSharedKeyCredential GetNewSharedKeyCredentials()
...
                 BlobName = blobName,
                 Protocol = SasProtocol.None,
                 StartsOn = Recording.UtcNow.AddHours(-1),
                 ExpiresOn = Recording.UtcNow.AddHours(+1),
-                Permissions = new BlobSasPermissions { Read = true, Add = true, Create = true, Write = true, Delete = true }.ToString(),
                 IPRange = new SasIPRange(IPAddress.None, IPAddress.None)
-            }.ToSasQueryParameters(sharedKeyCredentials ?? GetNewSharedKeyCredentials());
+            };
+            builder.SetPermissions(
+                BlobContainerSasPermissions.Read |
```
Comment (tg-msft) [inline sdk/storage/Azure.Storage.Blobs/tests/BlobTestBase.cs]: This was `BlobSasPermissions` but you're using `BlobContainerSasPermissions` here?

File: sdk/storage/Azure.Storage.Blobs/tests/BlobTestBase.cs [hunk 5]
```diff
@@ -246,75 +246,109 @@ public StorageSharedKeyCredential GetNewSharedKeyCredentials()
...
                 BlobName = blobName,
                 Protocol = SasProtocol.None,
                 StartsOn = Recording.UtcNow.AddHours(-1),
                 ExpiresOn = Recording.UtcNow.AddHours(+1),
-                Permissions = new BlobSasPermissions { Read = true, Add = true, Create = true, Write = true, Delete = true }.ToString(),
                 IPRange = new SasIPRange(IPAddress.None, IPAddress.None)
-            }.ToSasQueryParameters(userDelegationKey, accountName);
+            };
+            builder.SetPermissions(
+                BlobContainerSasPermissions.Read |
```
Comment (tg-msft) [inline sdk/storage/Azure.Storage.Blobs/tests/BlobTestBase.cs]: Ditto

File: sdk/storage/Azure.Storage.Files/src/Sas/FileSasBuilder.cs [hunk 6]
```diff
@@ -48,10 +49,10 @@ public class FileSasBuilder
         /// user is restricted to operations allowed by the permissions. This
         /// field must be omitted if it has been specified in an associated
         /// stored access policy.  The <see cref="FileSasPermissions"/>
-        /// and <see cref="ShareSasPermissions"/>
+        /// , <see cref="ShareSasPermissions"/>, or <see cref="FileAccountSasPermissions"/>
```
Comment (tg-msft) [inline sdk/storage/Azure.Storage.Files/src/Sas/FileSasBuilder.cs]: Nit - Move the comma back to the previous line

File: sdk/storage/Azure.Storage.Files/tests/FileTestBase.cs [hunk 7]
```diff
@@ -123,39 +123,48 @@ public StorageSharedKeyCredential GetNewSharedKeyCredentials()
...
                 Protocol = SasProtocol.None,
                 Services = AccountSasServices.Files,
                 ResourceTypes = AccountSasResourceTypes.Container,
                 StartsOn = Recording.UtcNow.AddHours(-1),
                 ExpiresOn = Recording.UtcNow.AddHours(+1),
-                Permissions = new FileAccountSasPermissions { Create = true, Delete = true }.ToString(),
                 IPRange = new SasIPRange(IPAddress.None, IPAddress.None)
-            }.ToSasQueryParameters(sharedKeyCredentials);
+            };
+            builder.SetPermissions(AccountSasPermissions.Create | AccountSasPermissions.Delete);
```
Comment (tg-msft) [inline sdk/storage/Azure.Storage.Files/tests/FileTestBase.cs]: You had `FileAccountSasPermissions` but changed it to `AccountSasPermissions`

File: sdk/storage/Azure.Storage.Queues/tests/QueueSasBuilderTests.cs [hunk 8]
```diff
@@ -90,11 +90,11 @@ private QueueSasBuilder BuildQueueSasBuilder(TestConstants constants, string que
                 Protocol = constants.Sas.Protocol,
                 StartsOn = constants.Sas.StartTime,
                 ExpiresOn = constants.Sas.ExpiryTime,
-                Permissions = Permissions,
                 IPRange = constants.Sas.IPRange,
                 Identifier = constants.Sas.Identifier,
                 QueueName = queueName,
             };
+            queueSasBuilder.SetPermissions(QueueSasPermissions.All);
```
Comment (tg-msft) [inline sdk/storage/Azure.Storage.Queues/tests/QueueSasBuilderTests.cs]: Why not pass the `Permissions` string here?

File: sdk/storage/Azure.Storage.Queues/tests/QueueTestBase.cs [hunk 9]
```diff
@@ -136,27 +136,40 @@ public StorageSharedKeyCredential GetNewSharedKeyCredentials()
...
                 Services = AccountSasServices.Queues,
                 ResourceTypes = AccountSasResourceTypes.Container,
                 StartsOn = Recording.UtcNow.AddHours(-1),
                 ExpiresOn = Recording.UtcNow.AddHours(+1),
-                Permissions = new QueueAccountSasPermissions { Read = true, Write = true, Update = true, Process = true, Add = true, Delete = true, List = true }.ToString(),
                 IPRange = new SasIPRange(IPAddress.None, IPAddress.None)
-            }.ToSasQueryParameters(sharedKeyCredentials);
+            };
+            builder.SetPermissions(
+                AccountSasPermissions.Read |
```
Comment (tg-msft) [inline sdk/storage/Azure.Storage.Queues/tests/QueueTestBase.cs]: QueueAccountSasPermissions

**Review by tg-msft — APPROVED** [review]

File: sdk/storage/Azure.Storage.Queues/tests/QueueTestBase.cs (hunk 9 above)
Comment (JoshLove-msft) [reply sdk/storage/Azure.Storage.Queues/tests/QueueTestBase.cs]: It is using an AccountSasBuilder though, so I think it was incorrect before.

File: sdk/storage/Azure.Storage.Files/tests/FileTestBase.cs (hunk 7 above)
Comment (JoshLove-msft) [reply sdk/storage/Azure.Storage.Files/tests/FileTestBase.cs]: It is using an AccountSasBuilder though, so I think it was incorrect before.

File: sdk/storage/Azure.Storage.Blobs/tests/BlobTestBase.cs (hunk 3 above)
Comment (JoshLove-msft) [reply sdk/storage/Azure.Storage.Blobs/tests/BlobTestBase.cs]: It is using an AccountSasBuilder though, so I think it was incorrect before.

File: sdk/storage/Azure.Storage.Blobs/tests/BlobTestBase.cs (hunk 4 above)
Comment (JoshLove-msft) [reply sdk/storage/Azure.Storage.Blobs/tests/BlobTestBase.cs]: Fixed

File: sdk/storage/Azure.Storage.Blobs/tests/BlobTestBase.cs (hunk 5 above)
Comment (JoshLove-msft) [reply sdk/storage/Azure.Storage.Blobs/tests/BlobTestBase.cs]: Fixed

**Review by amnguye — APPROVED** [review]

**Pull request merged**